src/Hedgehog/ListGen.fs

```diff
@@ -0,0 +1,16 @@
+﻿namespace Hedgehog
+
+module ListGen =
+
+    let traverse (f: 'a -> Gen<'b>) (ma: list<'a>) : Gen<list<'b>> =
```

Comment:
Out of curiosity, since this returns a `Gen<list<'b>>`, shouldn't the module be named `GenList`?

Reply:
This would be more consistent with `src/Hedgehog/Tuple.fs` in #247

Reply:
> Out of curiosity, since this returns a `Gen<list<'b>>`, shouldn't the module be named `GenList`?

Great question.  I always have to think carefully about which module name I think is correct.

Functions are typically contained in modules named after their input type.  For example, the [function `isSome`](https://fsharp.github.io/fsharp-core-docs/reference/fsharp-core-optionmodule.html#isSome) with type `'a option -> bool` is in a module named `Option` instead of one named `Bool`.  In addition to being typical, I think this is also the correct naming convention.  Given an instance of type `Foo`, a programmer knows they can type `Foo.` in their IDE and get suggestions about what they could do with their instance.

Of course `traverse` doesn't have `Gen<'a> list` (or `'a Gen list` when consistently using the postfix notation for generic types) as input.  Nor does it have `Gen<'a list>` (or `'a list Gen`) as input.  However, `traverse` should be in the same module as `sequence`, which can be implemented as `traverse id`.  The lone input to `sequence` has type `Gen<'a> list` (or `'a Gen list`).  Hence the module name `ListGen`.

_\*reflects on what was just written\*_

I just realized that this naming convention also depends on a preference for the generic type notation.  If the prefix notation is preferred, then my argument above says the module should be named `ListGen`, but if the postfix notation is preferred, then my argument above says the the module name should be `GenList`.

I prefer the prefix notation.  I think this is because I primarily used Java in school for ten years and then C# professionally for four years before picking up F#.

_\*more thinking\*_

The postfix generic type syntax is only available for generic types with a single type argument, right?  If so, then I think the module naming convention should be based on the prefix syntax.  Consider the nested type `Either<Map<'k, 'v, 'e>`.  It is a functor in `'v` (among other things).  In what module should functions for it like `map` exist?  Since the only way to express this type is with the prefix notation, I think the functions should be in a module named `EitherMap`.  Then for consistency, I think all nested types should have their functions contained a module matching the name of their type when expressed using the prefix notation.

> This would be more consistent with `src/Hedgehog/Tuple.fs` in #247

For reference, that module and its functions are [here](https://github.com/hedgehogqa/fsharp-hedgehog/pull/247/files#diff-0498fde986354191be905aa346ef56e4f314b6b38e18de8287148496ed061d31R12-R19).

I think `ListGen` and `GenTuple` are consistently named.  In general, I think you correctly named that module `GenTuple` because the outer type is `Gen<>` and the inner type is a tuple.

However, it is not just any tuple.  It is specifically a pair.  I just that you change the name of that module to `GenPair`.  I will make the same comment there for completeness.

Reply:
> ... because the outer type is Gen<> and the inner type is a tuple.

That's what I was getting at. The same is true here, where `Gen<>` is the outer type, and `List<>` is an inner type. If one prefers lighter weight generic syntax, it could be written as `Gen (List a)`. I would argue for naming based on an "outside-in" approach, where this PR uses "inside-out". I'm not sure if there's an established naming convention for this. I'm sure Haskell-ers have something, since they must deal with this more.

Module functions can also be grouped on the return value, lest we forget about `val constant : 'a -> Gen<'a>`. No one would argue that isn't a `Gen` function. Since `traverse` and `sequence` do the functor-flip, but both result in a `Gen<List<...>>` I would argue that the return type is more significant to the module they're contained in.

I'm genuinely curious if an established naming convention exists though, as I'd like to follow the principal of least surprise here.

Reply:
I'd also be totally fine with putting a `List` module under `Gen`, so accessing these functions would be `Gen.List.traverse`

Reply:
To summarize, there are three things to decide.
1. Outside-in approach vs inside-out approach
2. Prefix or postfix notation for generic types
3. Based on the type of an input or the output

> I would argue for naming based on an "outside-in" approach, where this PR uses "inside-out".

Me too.  I also want to name the modules based on the outside-in approach.

> > ... because the outer type is Gen<> and the inner type is a tuple.
> 
> `Gen<>` is the outer type, and `List<>` is an inner type [of the return type of `traverse`].

I want to name the modules based on the prefix notation of the generic type.  That quote makes me think that you do as well.

> Module functions can also be grouped on the return value, lest we forget about `val constant : 'a -> Gen<'a>`. No one would argue that isn't a `Gen` function. Since `traverse` and `sequence` do the functor-flip, but both result in a `Gen<List<...>>` I would argue that the return type is more significant to the module they're contained in.
> 
> I'm genuinely curious if an established naming convention exists though, as I'd like to follow the principle of least surprise here.

I think the only thing we disagree on there is whether the name of a module should match the type of an input or the output of the functions it contains.

I certainly admit that I want to put `traverse` in a module named `ListGen` even though `'a Gen list` isn't actually a type of one of its inputs.

I think your example of `constant` is misleading.  There are many functions merely meant to convert between types in a sense that is primarily syntactic.  A great example of this is `Seq.toList`.  Sometimes functions exist in both directions, such as `List.ofSeq`.  Sometimes only one of these to functions exists.  For your example, there is no containing type, so it is not even clear what name would be for the module containing the other function.  Sometimes one of these two functions is missing because F# makes it difficult to create cyclic dependencies.

In summary, there are many functions of the form `A.toB` and `B.ofA` both of type `'a -> 'b`.  My impression is that there are more such functions of the form `A.toB`.

> I'm not sure if there's an established naming convention for this.

Because there exists functions of both forms `A.toB` and `B.ofA`, any convention is not strictly followed.

When you questioned my naming, I was surprised to realize that there is more sublity than I had previously noticed and that I couldn't find any examples to justify my naming convention.

Here is one though, which is [the post by Scott Wlaschin about `traverse` and `sequence`](https://fsharpforfunandprofit.com/posts/elevated-world-4).  Specifically consider [this code](https://fsharpforfunandprofit.com/posts/elevated-world-4/#applicative-vs-monadic-versions-of-traverse:~:text=module%20List%20%3D,let%20rec%20traverseResultA%20f%20list%20%3D).
```fsharp
module List =

    /// Map a Result producing function over a list to get a new Result 
    /// using applicative style
    /// ('a -> Result<'b>) -> 'a list -> Result<'b list>
    let rec traverseResultA f list = [omitting implementation]
```

Scott's naming convention is to put `traverse` in a module named `List` to primarily match the last input type.  Of course it needs further distinguishing, so he appends `Result` to the function name.  (The `A` is orthogonal to our discussion.)  So Scott is primarily emphasizing `List` and secondarily emphasizing `Result`.  I agree with that order of emphasis.  I just think it is better to move `Result` up to a module and simplify the function name to just `traverse`.

> I'd also be totally fine with putting a `List` module under `Gen`, so accessing these functions would be `Gen.List.traverse`

This is a good idea, but of course I think it should be `List.Gen.traverse`.  This would also be more consistent with the fact that the last input of traverse is `'a list` (with `'a = _ Gen`).

> I'm sure Haskell-ers have something, since they must deal with this more.

This is not an issue for them.  Higher-kinded types allow them to simply write `map` instead of `Gen.map`, `List.map`, etc.